# Cryptography Playground
## Binary Conversion

In [1]:
# Using bin()
num = 5
print(bin(num))           # '0b101' (0b means binary)
print(bin(num)[2:])       # '101' (remove 0b prefix)

# With formatting
print(f"{num:b}")         # '101'
print(f"{num:08b}")       # '00000101' (8 bits, padded with zeros)

0b101
101
101
00000101


In [2]:
# Single character
letter = 'A'
print(f"Letter = {letter}")
print(f"{letter} to number conversion = {ord(letter)}") # You can only XOR the number (the numeric value) that ord() gives you, not the character itself.
print(f"Get binary from number = {bin(ord(letter))}") # This is giving you the binary of the resulting number from ord()
print(f"Same binary but different = {ord(letter):08b}") # This is same as above but :08b is asking for a specific format

print(f"\n{'###' * 20}\n") #################

# Whole string, the "1" works because ord() is called but does not represent the binary for the integer 1
text = "Aa1"
for char in text:
    print(f"{char} = {ord(char):08b}")

print(f"\n{'###' * 20}\n") #################

# STRING AND INTEGER BINARIES ARE 2 DIFFERENT THINGS
print(f'String 1 = {ord('1'):08b}') # ord() converts the '1' character to an integer - but not the 1 integer.
print(f'Integer 1 = {1:08b}')
print(f"ord('1') result = {ord('1')}")  # 49 (code point of character '1')


Letter = A
A to number conversion = 65
Get binary from number = 0b1000001
Same binary but different = 01000001

############################################################

A = 01000001
a = 01100001
1 = 00110001

############################################################

String 1 = 00110001
Integer 1 = 00000001
ord('1') result = 49


In [3]:
text1 = "HELLO"
binary1 = ' '.join(f"{ord(char):08b}" for char in text1)
print(f"text1 = {binary1}")

text2 = "hello"
binary2 = ' '.join(f"{ord(char):08b}" for char in text2)
print(f"text1 = {binary2}")


text1 = 01001000 01000101 01001100 01001100 01001111
text1 = 01101000 01100101 01101100 01101100 01101111


In [4]:
# WORKS BECAUSE INTERGER
print(f"{1:08b}")

# OR
ABC = 1
print(f"{ABC:08b}")

# OR IF YOU CALL ord()
print(f"{ord('1'):08b}")
# BUT the '1' character (as a string) is different to the integer 1

print(1)           # 1 (integer)
print(ord('1'))    # 49 (code point of character '1')

print(f"{1:08b}")        # 00000001 (binary of 1)
print(f"{49:08b}")       # 00110001 (binary of 49)
print(f"{ord('1'):08b}") # 00110001 (same as above!)

00000001
00000001
00110001
1
49
00000001
00110001
00110001


In [5]:
# DOESN'T WORK BECAUSE STRING NOT CALLED WITH ord()
# print(f"{'1':08b}")

# SAME HERE, NO ord() ON STRING
# ABC = '1'
# print(f"{ABC:08b}")

### Binary Conversion Summary
- `print(f"{integer:08b}")`
- Can only convert integers to binary so use ord() on characters/strings to convert them to their integer equivalent. But running ord() on '2' does not convert the integer 2 to binary. It converts the character '2' to binary. 
- So be cautious of the datatype.. '5' the string, and it's ord() is not representative of the integer 5. The character '5' will have a different binary to the integer 5. So make sure you're working on the right datatype, and have used ord() appropriately.
- The catch is that ord() only works on single characters not on strings so you have to use a for loop on the string. (Seen above)
- If you can't convert characters to binary than you absolutely can't XOR characters directly to use `^` which is XOR in python, you need to run ord() on characters or be sure that you're working with integers already.

## XORing
So once you've made everything an integer, so that you can make it into a binary, then you can start XORing things together with `^`. Although, what's presented to you is the symbol for the integer, in memory it's already binary so you're not actually converting it to binary or making it into a binary, it's already in binary form, just presented to you as a int. 

In [6]:
# Direct XOR with integers as integers
a = 5
b = 3
result = a ^ b

print(f"As integers: {a} ^ {b} = {result}")

print("")

print (f"As binary: {a:08b} ^ {b:08b} = {result:08b}") # What XOR cares about

print("")

print(f"5 as a binary = {5:08b}")
print(f"3 as a binary = {3:08b}")
print(f"6 as a binary = {6:08b}")






As integers: 5 ^ 3 = 6

As binary: 00000101 ^ 00000011 = 00000110

5 as a binary = 00000101
3 as a binary = 00000011
6 as a binary = 00000110


Strictly speaking when we 5 ^ 3, Python isn't converting either integer to binary to then XOR them. They're already stored in binary in the computer's memory.

What we write (decimal for humans):
- `5 ^ 3`

What the computer sees (binary in hardware):
- `101 ^ 011 = 110`

What result gets displayed back to you (decimal):
- `6` but it's really `00000110`

Behind the scenes:
- `5` is stored as `101` in memory
- `3` is stored as `011` in memory  
- The `^` operator XORs them **bit by bit** directly


When you see 5, that's just Python's way of displaying the binary value 101 in a format humans prefer (decimal).

In [8]:
result = 0 ^ 0
print(result)

result = 0 ^ 1
print(result)

result = 1 ^ 0
print(result)

result = 1 ^ 1
print(result)


0
1
1
0


FAMILIAR?

| A | B | A ^ B |
|---|---|-------|
| 0 | 0 |   0   |
| 0 | 1 |   1   |
| 1 | 0 |   1   |
| 1 | 1 |   0   |

Where A is one int compared to the equivalently indexed int in B